### DISCUSSION TITLE
Amazing project

### DISCUSSION THREAD

**Discussion by FoolHen** [rerdavies/pipedal#139] [General]
I recently found this project and I'm really impressed with how well it works. I recently got an RPi 4 and was thinking about creating a pedal with it, and looking for some options I stumbled upon this.

After trying it out for a bit I can say that it's really well done and is way more elaborate than what I thought was possible. I was surprised that I only could find two videos about it on Youtube. I would say that this has huge potential, especially with the support for NAM models. If someone with a following makes a video I'm sure this will gain a lot of attention, even more if its built into a DIY pedal.

I've been trying it with my RPi 4 and a simple USB audio interface from an old headset and it works really well, although with noise and some latency as it is not that good. My idea is to create a pedal with this, using a good audio interface (probably the Hifiberry DAC+ ADC) and adding the jack inputs, a foot switch and some foot buttons and convert into a mini multi-fx, muti-purpose pedal.

I have a couple of questions that I would like to throw, as I'm a noob about linux and audio. First, I saw that there are some linux-based OS's like Elk Audio OS that focus on improving the audio performance. I believe that PiPedal would be compatible. Do you think that would offer an improvement for PiPedal or do you believe that it's not necessary? 

Another question that I have is related to MIDI controls. As I said I was thinking on adding buttons to the RPi in order to set them to do specific actions. How simple is that? Would I have to create a script to read the buttons connected to the GPIO pins and convert it to MIDI events?

This also brings me the question, how is your set up @rerdavies ? Are you running something similar to what I have in mind?

Lastly, and more importantly, huge thanks for this amazing project. I'm really happy that I found this, and would love to contribute to the project even though my knowledge about this topic is quite limited. I'm a web/mobile developer so I would be able to help on that side if needed if I have some extra free time. Thank you and looking forward to your response.

**Comment by adbrt** [comment]
@FoolHen I managed to simulate MIDI commands, this is how it can be done: https://github.com/rerdavies/pipedal/discussions/141

BTW I am using Behringer UCG102 USB interface and I am getting 4.9ms measured latency, 48kHz 64 samples 2 buffers. I measured it by using two microphones, one directed to the headphones and other directed to the guitar, so it is total guitar-to-speaker latency. This is as low as it can get with this interface. I noticed that turning on the web interface can occasionaly create some dropouts in this configuration, which can be fixed by using 3 buffers. But I guess that with GPIO control it would no longer be an issue.

**Reply by FoolHen** [reply]
Oh awesome, thank you. I will try that when I install the buttons. I was finally able to try this properly with an audio interface for the rpi and I've noticed that for the NAM plugin to work it has to use at least 3 buffers and minimum 64 samples. Anything lower makes glitchy noises. I haven't tested the latency yet tho, I don't have the setup for that right now.

**Comment by 38github** [comment]
I am so grateful to also have found PiPedal.

I tried MODEP with its Patchbox OS on a RPi4 and it generated xruns and used a lot of CPU. Then I installed a regular Debian distro, did the audio group, threadirqs and mitigations=off and installed PiPedal. I can now use 64 samples without a problem and a NAM Standard WaveNet model, mono delay, parametric EQ and a speaker IR combination uses ~30% CPU (Pipedal indicator) and no single xrun even after several hours.

I think that MODEP is nice looking and fun to play around with but I prefer high stability over that. One day I will try a MIDI footswitch pedal, create a few different setups and see how well PiPedal works with that.

**Comment by rerdavies** [comment]
> I've been trying it with my RPi 4 and a simple USB audio interface from an old headset and it works really well, although with noise and some latency as it is not that good. My idea is to create a pedal with this, using a good audio interface (probably the Hifiberry DAC+ ADC) and adding the jack inputs, a foot switch and some foot buttons and convert into a mini multi-fx, muti-purpose pedal.

I have not used the Hifiberry DAC+, but I see a number of pain points.  

- A suspicious lack of medium sized capacitors, which makes me think it's going to have poor S/N ratio.
- Getting instrument-level 1/4" jacks into line-level RCA jacks isn't going to be fun or straightfoward. You'll probably need a mixer to do it.

Personally, I can't recommend the MOTU M2, which connects via USb, highly enough.  The DACs/ADCs actually are audiophile quality, a good 18-20 bits of signal above the noise floor, which is exceptional. I've even seen an audiophile magazine recommending the M2 in favor of audiophile ADC/DACs that costs hundreds more.  And I like the cabling simplicity and the ability to trim input levels using knobs on the front of the M2. 

A dream future project of mine: 3d print a case that holds both my PI, (and my SD DRIVE) and the M2 in a single box. 


> First, I saw that there are some linux-based OS's like Elk Audio OS that focus on improving the audio performance. I believe that PiPedal would be compatible. 

PiPedal would not be remotely compatible with Elk Audio OS. It's actually quite an awkward platform to develop for, with not a huge following. 

And it's not immediately clear to me that Elk Audio OS is going to provide better latency than a Linux RT_PREEMPT kernel. Some of the Raspberry Pi OS drivers seem to be missing RT patches. So heavy file I/O while doing audio-synthesis may cause dropouts.  (This may only happen with SDCARD drivers, and not with USB drives). So it's not a perfect real-time OS. But it's surprisingly close.

> Another question that I have is related to MIDI controls. As I said I was thinking on adding buttons to the RPi in order to set them to do specific actions. How simple is that? Would I have to create a script to read the buttons connected to the GPIO pins and convert it to MIDI events?

Yes. PiPedal does not support that.  I personally use the smallest midi USB controller I could find (6 pads) to do what I need.

> This also brings me the question, how is your set up @rerdavies ? Are you running something similar to what I have in mind?

My setup:

- Motu M2 USb audio device.
- Pipedal.
- Cheap roland USB pad controller for bypassing chorus/flanger effects, depending on the patch, and patch selection.
- A 100W-ish speaker (A "bazillion watt" class D), which also has a bluetooth connection for streaming audio from my phone.

It irritates me slightly, as a point of honor, that I have to rely on my speaker's bluetooth channel, instead of the second input from my M2, or maybe the bluetooth channel on my PI. Both features requests that are pretty far down the list at present, but certainly ideas for future features. In theory, you could route the second input channel straight through using route splits. But it seems like there should be an easier way to do that in PiPedal. And besides, my speaker has a mixer, also controllable from my phone, so it hasn't been a personal priority.

> huge thanks for this amazing project... I'm a web/mobile developer so I would be able to help on that side if needed..

A huge you're welcome. If you have experience with React framework, your collaboration would be hugely welcome.  I have a couple of mid-sized features I would love help with. Drop me a note: rerdavies at gmail.com. 

Most pressing issue I would love help with: move from react framework (which is no longer supported) to the Vite toolchain (which is why react framework is no longer supported) for the website.  Pretty straightforward, mostly moving files around into the right places, and a small glue changes to hook the root page into the framework index.htm page. Probably a couple of days work.
